File: audiocloud-api/src/driver.rs
//! Types used to communicate with the driver

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::change::{PlayId, RenderId};
use crate::instance::{DesiredInstancePlayState, InstancePlayState};
use crate::model::MultiChannelValue;
use crate::newtypes::{ParameterId, ReportId};

#[derive(PartialEq, Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "snake_case")]
pub enum InstanceDriverCommand {
  CheckConnection,
  Stop,
  Play { play_id: PlayId },
  Render { length: f64, render_id: RenderId },
  Rewind { to: f64 },
  SetParameters(HashMap<ParameterId, MultiChannelValue>),
}

#[derive(PartialEq, Serialize, Deserialize, Clone, Debug, Error)]
#[serde(rename_all = "snake_case")]
pub enum InstanceDriverError {
  #[error("Parameter {parameter} does not exist")]
  ParameterDoesNotExist { parameter: String },

  #[error("Media is not present, can't play or rewind")]
  MediaNotPresent,

  #[error("Driver can't guarantee that playback won't be interrupted")]
  NotInterruptable,

  #[error("Remote call failed: {error}")]
  RPC { error: String },
}

#[derive(PartialEq, Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "snake_case")]
pub enum InstanceDriverEvent {
  /// Sent when the driver has started
  Started,

  /// If an I/O error happened during communication with device
  IOError { error: String },

  /// Driver lost connection to the hardware
  ConnectionLost,

  /// Driver connected to the hardware
  Connected,

  /// Received metering updates from the hardware
  Metering { meters: HashMap<ReportId, MultiChannelValue> },

  /// Playing; media current position reported
  PlayState {
    desired: DesiredInstancePlayState,
    current: InstancePlayState,
    media:   Option<f64>,
  },
}

File: js-api/src/driver.ts
import { Static, Type } from "@sinclair/typebox";
import Option from "./utils/option";
import { PlayId, RenderId } from "./change";
import { DesiredInstancePlayState, InstancePlayState } from "./instance";
import { ParameterId } from "./new_types";
import { InstanceReports } from "./session";
import { MultiChannelValue } from "./model";

export const InstanceDriverCommand = Type.Union([
    Type.Literal("check_connection"),
    Type.Literal("stop"),
    Type.Object({
        "play":                         Type.Object({
            "play_id":                  PlayId
        })
    }),
    Type.Object({
        "render":                       Type.Object({
            "length":                   Type.Number(),
            "render_id":                RenderId
        })
    }),
    Type.Object({
        "rewind":                       Type.Object({
            "to":                       Type.Number()
        })
    }),
    Type.Object({ "set_parameters":     Type.Record(ParameterId, MultiChannelValue) }),
])
export type InstanceDriverCommand = Static<typeof InstanceDriverCommand>

export const InstanceDriverError = Type.Union([
    Type.Object({
        "parameter_does_not_exist":     Type.Object({
            "parameter":                Type.String()
        })
    }),
    Type.Literal("media_not_present"),
    Type.Literal("not_interruptable"),
    Type.Object({
        "rpc":                          Type.Object({
            "error":                    Type.String()
        })
    })
])
export type InstanceDriverError = Static<typeof InstanceDriverError>

export const InstanceDriverEvent = Type.Union([
    /// Sent when the driver has started
    Type.Literal("started"),

    /// If an I/O error happened during communication with device
    Type.Object({ "io_error":           Type.Object({ "error": Type.String() }) }),

    /// Driver lost connection to the hardware
    Type.Literal("connection_lost"),

    /// Driver connected to the hardware
    Type.Literal("connected"),

    /// Received metering updates from the hardware
    Type.Object({ "reports":            Type.Object({ "reports": InstanceReports }) }),

    /// Playing; media current position reported
    Type.Object({
        "play_state":                   Type.Object({
            "desired":                  DesiredInstancePlayState,
            "current":                  InstancePlayState,
            "media":                    Option(Type.Number()),
        })
    }),
])
export type InstanceDriverEvent = Static<typeof InstanceDriverEvent>